File: android/app/src/main/java/dev/rahier/pouleparty/ui/components/GameStartCountdownOverlay.kt
package dev.rahier.pouleparty.ui.components

import androidx.compose.animation.*
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.outlined.ContentCopy
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.rahier.pouleparty.AppConstants
import dev.rahier.pouleparty.R
import dev.rahier.pouleparty.ui.theme.*
import kotlinx.coroutines.delay
import java.util.Date

@Composable
fun GameStartCountdownOverlay(
    countdownNumber: Int?,
    countdownText: String?
) {
    AnimatedVisibility(
        visible = countdownNumber != null || countdownText != null,
        enter = fadeIn(animationSpec = tween(200)),
        exit = fadeOut(animationSpec = tween(300))
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.4f)),
            contentAlignment = Alignment.Center
        ) {
            if (countdownNumber != null) {
                Text(
                    text = "$countdownNumber",
                    style = gameboyStyle(80),
                    color = CROrange,
                    modifier = Modifier.neonGlow(CROrange, NeonGlowIntensity.INTENSE)
                )
            } else if (countdownText != null) {
                Text(
                    text = countdownText,
                    style = bangerStyle(48),
                    color = CROrange,
                    modifier = Modifier.neonGlow(CROrange, NeonGlowIntensity.MEDIUM)
                )
            }
        }
    }
}

/** Role identifies which header copy + launch-button affordance the
 *  pre-game overlay should display. Mirrors iOS `GameRole`. */
enum class PreGameRole { CHICKEN, HUNTER, GAME_MASTER }

/**
 * Full-screen "lobby" overlay shown on every active map (chicken /
 * hunter / GameMaster) **before** the game starts. Three flavours:
 *
 *  - Auto-start (default): countdown to `targetDate`. The overlay
 *    self-hides once the 3-2-1-GO! `GameStartCountdownOverlay` takes
 *    over.
 *  - Manual-start launcher (`isManualStart == true` && role is
 *    chicken or game master): big LAUNCH button. Calls `onLaunchTapped`,
 *    shows spinner + "Launching…" while the callable is in flight,
 *    surfaces `launchErrorMessage` as an AlertDialog.
 *  - Manual-start waiter (`isManualStart == true` && role is hunter):
 *    passive "Waiting for the chicken to launch" message.
 *
 * Replaces the previous `ReadyToLaunchOverlay` (deleted with this
 * change) so all pre-game UI lives in one component.
 */
@Composable
fun PreGameOverlay(
    role: PreGameRole,
    gameModTitle: String,
    gameCode: String?,
    targetDate: Date,
    nowDate: Date,
    connectedHunters: Int = 0,
    onCancelGame: (() -> Unit)? = null,
    isManualStart: Boolean = false,
    isLaunching: Boolean = false,
    launchErrorMessage: String? = null,
    onLaunchTapped: (() -> Unit)? = null,
    onLaunchErrorDismissed: (() -> Unit)? = null,
) {
    val context = LocalContext.current
    val secondsRemaining = maxOf(0L, (targetDate.time - nowDate.time) / 1000)
    val days = secondsRemaining / 86400
    val hours = (secondsRemaining % 86400) / 3600
    val minutes = (secondsRemaining % 3600) / 60
    val seconds = secondsRemaining % 60
    val formattedTime = when {
        days > 0 -> String.format("%dj %02d:%02d:%02d", days, hours, minutes, seconds)
        hours > 0 -> String.format("%d:%02d:%02d", hours, minutes, seconds)
        else -> String.format("%d:%02d", minutes, seconds)
    }
    val timerFontSize = when {
        days > 0 -> 30
        hours > 0 -> 38
        else -> 48
    }
    var codeCopied by remember { mutableStateOf(false) }

    LaunchedEffect(codeCopied) {
        if (codeCopied) {
            delay(1500)
            codeCopied = false
        }
    }

    // Manual-start stays visible until the launcher taps LAUNCH (the
    // parent unmounts the overlay by flipping out of READY_TO_LAUNCH).
    // Auto-start hides once the 3-2-1 takes over.
    val isVisible = isManualStart ||
        secondsRemaining > AppConstants.COUNTDOWN_THRESHOLD_SECONDS.toLong()

    val headerRes = when (role) {
        PreGameRole.CHICKEN -> R.string.you_are_chicken
        PreGameRole.HUNTER -> R.string.you_are_hunter
        PreGameRole.GAME_MASTER -> R.string.you_are_gamemaster
    }
    val isLauncherRole = role == PreGameRole.CHICKEN || role == PreGameRole.GAME_MASTER

    AnimatedVisibility(
        visible = isVisible,
        exit = fadeOut(animationSpec = tween(500))
    ) {
        // Manual-start uses Top alignment + a flexible Spacer between
        // info and launch so the launch button + cancel land in the
        // lower half (~3/4 height) — natural thumb reach instead of a
        // centered-card feel. Auto-start keeps the centered layout.
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.7f)),
            contentAlignment = if (isManualStart) Alignment.TopCenter else Alignment.Center
        ) {
            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(24.dp),
                modifier = if (isManualStart) {
                    Modifier.fillMaxSize().padding(horizontal = 32.dp, vertical = 32.dp)
                } else {
                    Modifier.padding(32.dp)
                },
            ) {
                Text(
                    text = stringResource(headerRes),
                    style = bangerStyle(32),
                    color = Color.White
                )

                Text(
                    text = gameModTitle,
                    fontSize = 16.sp,
                    color = Color.White.copy(alpha = 0.7f)
                )

                if (gameCode != null) {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                        modifier = Modifier
                            .padding(top = 8.dp)
                            .clickable {
                                copyToClipboard(context, "Game Code", gameCode)
                                codeCopied = true
                            }
                    ) {
                        Text(
                            text = if (codeCopied) stringResource(R.string.copied) else stringResource(R.string.tap_to_copy),
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                        Row(
                            verticalAlignment = Alignment.CenterVertically,
                            horizontalArrangement = Arrangement.spacedBy(12.dp)
                        ) {
                            Text(
                                text = gameCode,
                                style = gameboyStyle(28),
                                color = Color.White
                            )
                            Icon(
                                imageVector = if (codeCopied) Icons.Default.Check else Icons.Outlined.ContentCopy,
                                contentDescription = stringResource(R.string.copy_game_code),
                                tint = if (codeCopied) Success else Color.White.copy(alpha = 0.6f)
                            )
                        }
                    }
                }

                Column(
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(6.dp),
                    modifier = Modifier.padding(top = 8.dp)
                ) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(6.dp)
                    ) {
                        Text(text = "🐔", fontSize = 14.sp)
                        Text(
                            text = "1",
                            style = gameboyStyle(16),
                            color = Color.White
                        )
                        Text(
                            text = stringResource(R.string.connected),
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                    }
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(6.dp)
                    ) {
                        Text(text = "🔍", fontSize = 14.sp)
                        Text(
                            text = "$connectedHunters",
                            style = gameboyStyle(16),
                            color = Color.White
                        )
                        Text(
                            text = stringResource(R.string.connected),
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                    }
                }

                if (isManualStart) {
                    // Flexible spacer pushes the launch section + cancel
                    // toward the lower half of the screen.
                    Spacer(Modifier.weight(1f).fillMaxWidth())
                    if (isLauncherRole && onLaunchTapped != null) {
                        Text(
                            text = "Ready when you are",
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                        Button(
                            onClick = onLaunchTapped,
                            enabled = !isLaunching,
                            modifier = Modifier
                                .fillMaxWidth()
                                .height(64.dp)
                                .background(
                                    brush = Brush.linearGradient(listOf(CROrange, CRPink)),
                                    shape = RoundedCornerShape(16.dp),
                                ),
                            shape = RoundedCornerShape(16.dp),
                            colors = ButtonDefaults.buttonColors(
                                containerColor = Color.Transparent,
                                disabledContainerColor = Color.Transparent,
                            ),
                        ) {
                            Row(
                                verticalAlignment = Alignment.CenterVertically,
                                horizontalArrangement = Arrangement.spacedBy(12.dp),
                            ) {
                                if (isLaunching) {
                                    CircularProgressIndicator(
                                        color = Color.White,
                                        strokeWidth = 3.dp,
                                        modifier = Modifier.size(20.dp),
                                    )
                                }
                                Text(
                                    text = if (isLaunching) "Launching…" else "LAUNCH GAME",
                                    color = Color.White,
                                    fontSize = 22.sp,
                                    fontWeight = FontWeight.Bold,
                                )
                            }
                        }
                    } else {
                        Text(
                            text = "Waiting for the chicken to launch",
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.7f),
                            textAlign = TextAlign.Center,
                        )
                        CircularProgressIndicator(
                            color = Color.White,
                            strokeWidth = 3.dp,
                            modifier = Modifier.size(32.dp),
                        )
                    }
                } else {
                    Column(
                        horizontalAlignment = Alignment.CenterHorizontally,
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                        modifier = Modifier.padding(top = 16.dp)
                    ) {
                        Text(
                            text = stringResource(R.string.game_starting_in),
                            fontSize = 14.sp,
                            color = Color.White.copy(alpha = 0.6f)
                        )
                        Text(
                            text = formattedTime,
                            style = gameboyStyle(timerFontSize),
                            color = CROrange
                        )
                    }
                }

                if (onCancelGame != null) {
                    TextButton(
                        onClick = onCancelGame,
                        modifier = Modifier.padding(top = 8.dp)
                    ) {
                        Text(
                            text = stringResource(R.string.cancel_game),
                            fontSize = 16.sp,
                            fontWeight = FontWeight.Medium,
                            color = Danger
                        )
                    }
                }
                if (isManualStart) {
                    // Lower bottom margin so the launch chunk sits at
                    // roughly 70-75% screen height, not glued to the
                    // bottom safe area. Mirrors the iOS layout shift.
                    Spacer(Modifier.weight(0.4f).fillMaxWidth())
                }
            }
        }
    }

    if (launchErrorMessage != null) {
        AlertDialog(
            onDismissRequest = { onLaunchErrorDismissed?.invoke() },
            title = { Text("Launch failed") },
            text = { Text(launchErrorMessage) },
            confirmButton = {
                TextButton(onClick = { onLaunchErrorDismissed?.invoke() }) { Text("OK") }
            },
        )
    }
}

File: ios/PouleParty/Components/GameStartCountdownOverlay.swift
//
//  GameStartCountdownOverlay.swift
//  PouleParty
//

import SwiftUI

struct GameStartCountdownOverlay: View {
    let countdownNumber: Int?
    let countdownText: String?

    var body: some View {
        ZStack {
            if let number = countdownNumber {
                Color.black.opacity(0.4)
                    .ignoresSafeArea()

                Text("\(number)")
                    .font(.gameboy(size: 80))
                    .foregroundStyle(Color.CROrange)
                    .neonGlow(.CROrange, intensity: .intense)
                    .id(number)
                    .transition(.scale.combined(with: .opacity))
                    .animation(.easeOut(duration: 0.3), value: number)
            } else if let text = countdownText {
                Color.black.opacity(0.4)
                    .ignoresSafeArea()

                BangerText(text, size: 48)
                    .foregroundStyle(Color.CROrange)
                    .transition(.scale.combined(with: .opacity))
            }
        }
        .animation(.easeInOut(duration: 0.3), value: countdownNumber)
        .animation(.easeInOut(duration: 0.3), value: countdownText)
    }
}

// MARK: - Pre-Game Overlay

/// Full-screen "lobby" overlay shown on every active map (chicken /
/// hunter / GameMaster) **before** the game starts. Three flavours:
///
/// - **Auto-start (default)**: shows the countdown to `targetDate`
///   (planned `timing.start`). When the countdown reaches the
///   `countdownThresholdSeconds` window the overlay self-hides so
///   the 3-2-1-GO! `GameStartCountdownOverlay` takes over.
/// - **Manual-start launcher** (`isManualStart == true` && role is
///   chicken or GameMaster): replaces the countdown copy with the big
///   LAUNCH button. Tapping calls `onLaunchTapped`. Shows a spinner +
///   "Launching…" while the callable is in flight, and an inline
///   alert via `launchErrorMessage`.
/// - **Manual-start waiter** (`isManualStart == true` && role is
///   hunter): passive "Waiting for the chicken to launch" message.
struct PreGameOverlay: View {
    let role: GameRole
    let gameModTitle: String
    let gameCode: String?
    let targetDate: Date
    let nowDate: Date
    var connectedHunters: Int = 0
    var onCancelGame: (() -> Void)? = nil
    /// PP-71: when true, the overlay shifts from "countdown to start"
    /// into the manual-launch mode (LAUNCH button for chicken/GM,
    /// passive waiter for hunters).
    var isManualStart: Bool = false
    var isLaunching: Bool = false
    var launchErrorMessage: String? = nil
    var onLaunchTapped: (() -> Void)? = nil
    var onLaunchErrorDismissed: (() -> Void)? = nil

    @State private var codeCopied = false

    private var secondsRemaining: Int {
        max(0, Int(targetDate.timeIntervalSince(nowDate)))
    }

    private var days: Int { secondsRemaining / 86400 }
    private var hours: Int { (secondsRemaining % 86400) / 3600 }

    private var formattedTime: String {
        let minutes = (secondsRemaining % 3600) / 60
        let seconds = secondsRemaining % 60
        if days > 0 {
            return String(format: "%dj %02d:%02d:%02d", days, hours, minutes, seconds)
        } else if hours > 0 {
            return String(format: "%d:%02d:%02d", hours, minutes, seconds)
        } else {
            return String(format: "%d:%02d", minutes, seconds)
        }
    }

    private var timerFontSize: CGFloat {
        if days > 0 { return 30 }
        if hours > 0 { return 38 }
        return 48
    }

    /// Hide the auto-start countdown overlay when the 3-2-1 takes
    /// over. The manual-start variant stays visible until the
    /// launcher taps LAUNCH (the parent un-mounts the overlay by
    /// flipping out of `.readyToLaunch`).
    private var isVisible: Bool {
        if isManualStart { return true }
        return secondsRemaining > Int(AppConstants.countdownThresholdSeconds)
    }

    private var isLauncherRole: Bool {
        role == .chicken || role == .gameMaster
    }

    private var headerTitle: String {
        switch role {
        case .chicken: return "You are the 🐔"
        case .hunter: return "You are the Hunter"
        case .gameMaster: return "You are the GameMaster 🦅"
        }
    }

    var body: some View {
        ZStack {
            if isVisible {
                Color.black.opacity(0.7)
                    .ignoresSafeArea()

                VStack(spacing: 24) {
                    BangerText(headerTitle, size: 32)
                        .foregroundStyle(.white)

                    Text(gameModTitle)
                        .font(.system(size: 16))
                        .foregroundStyle(.white.opacity(0.7))

                    if let code = gameCode {
                        Button {
                            UIPasteboard.general.string = code
                            withAnimation { codeCopied = true }
                            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) {
                                withAnimation { codeCopied = false }
                            }
                        } label: {
                            VStack(spacing: 8) {
                                Text(codeCopied ? "Copied!" : "Tap to copy:")
                                    .font(.system(size: 14))
                                    .foregroundStyle(.white.opacity(0.6))
                                HStack(spacing: 12) {
                                    Text(code)
                                        .font(.gameboy(size: 28))
                                        .foregroundStyle(.white)
                                    Image(systemName: codeCopied ? "checkmark" : "doc.on.doc")
                                        .foregroundStyle(codeCopied ? .green : .white.opacity(0.6))
                                        .contentTransition(.symbolEffect(.replace))
                                }
                            }
                        }
                        .buttonStyle(.plain)
                        .padding(.top, 8)
                    }

                    VStack(spacing: 6) {
                        HStack(spacing: 6) {
                            Text("🐔")
                                .font(.system(size: 14))
                            Text("1")
                                .font(.gameboy(size: 16))
                                .foregroundStyle(.white)
                            Text("connected")
                                .font(.system(size: 14))
                                .foregroundStyle(.white.opacity(0.6))
                        }
                        HStack(spacing: 6) {
                            Text("🔍")
                                .font(.system(size: 14))
                            Text("\(connectedHunters)")
                                .font(.gameboy(size: 16))
                                .foregroundStyle(.white)
                            Text("connected")
                                .font(.system(size: 14))
                                .foregroundStyle(.white.opacity(0.6))
                        }
                    }
                    .padding(.top, 8)

                    if isManualStart {
                        // Push the manual-launch CTA + Cancel button
                        // toward the lower portion of the screen
                        // (~3/4 height) so the launcher's thumb has a
                        // natural reach and the info up top reads as
                        // a header rather than a centered card.
                        Spacer(minLength: 24)
                            .frame(maxHeight: .infinity)
                        manualStartSection

                        if let onCancelGame {
                            Button {
                                onCancelGame()
                            } label: {
                                Text("Cancel game")
                                    .font(.system(size: 16, weight: .medium))
                                    .foregroundStyle(Color.danger)
                            }
                            .padding(.top, 4)
                        }
                        Spacer().frame(height: 60)
                    } else {
                        VStack(spacing: 8) {
                            Text("Game starts in")
                                .font(.system(size: 14))
                                .foregroundStyle(.white.opacity(0.6))
                            Text(formattedTime)
                                .font(.gameboy(size: timerFontSize))
                                .foregroundStyle(Color.CROrange)
                                .neonGlow(.CROrange, intensity: .medium)
                                .contentTransition(.numericText())
                                .animation(.linear(duration: 0.3), value: secondsRemaining)
                        }
                        .padding(.top, 16)

                        if let onCancelGame {
                            Button {
                                onCancelGame()
                            } label: {
                                Text("Cancel game")
                                    .font(.system(size: 16, weight: .medium))
                                    .foregroundStyle(Color.danger)
                            }
                            .padding(.top, 8)
                        }
                    }
                }
                .frame(maxWidth: .infinity, maxHeight: isManualStart ? .infinity : nil)
                .padding(32)
            }
        }
        .animation(.easeOut(duration: 0.4), value: isVisible)
        .alert(
            "Launch failed",
            isPresented: Binding(
                get: { launchErrorMessage != nil },
                set: { newValue in if !newValue { onLaunchErrorDismissed?() } }
            ),
            presenting: launchErrorMessage
        ) { _ in
            Button("OK", role: .cancel) { onLaunchErrorDismissed?() }
        } message: { message in
            Text(message)
        }
    }

    @ViewBuilder
    private var manualStartSection: some View {
        if isLauncherRole, let onLaunchTapped {
            VStack(spacing: 12) {
                Text("Ready when you are")
                    .font(.system(size: 14))
                    .foregroundStyle(.white.opacity(0.6))
                Button(action: onLaunchTapped) {
                    HStack(spacing: 12) {
                        if isLaunching {
                            ProgressView()
                                .progressViewStyle(.circular)
                                .tint(.white)
                        }
                        Text(isLaunching ? "Launching…" : "LAUNCH GAME")
                            .font(.title.bold())
                            .foregroundStyle(.white)
                    }
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 18)
                    .background(
                        LinearGradient(
                            colors: [Color.CROrange, Color.CRPink],
                            startPoint: .topLeading,
                            endPoint: .bottomTrailing
                        )
                    )
                    .clipShape(RoundedRectangle(cornerRadius: 16))
                }
                .disabled(isLaunching)
            }
        } else {
            VStack(spacing: 8) {
                Text("Waiting for the chicken to launch")
                    .font(.system(size: 14))
                    .foregroundStyle(.white.opacity(0.7))
                ProgressView()
                    .progressViewStyle(.circular)
                    .tint(.white)
                    .scaleEffect(1.2)
            }
        }
    }
}
